src/pybind/mgr/cephadm/serve.py

```diff
@@ -776,9 +776,6 @@ def _create_daemon(self,
                 if haspec.keepalived_container_image:
                     image = haspec.keepalived_container_image
 
-            cephadm_config, deps = self.mgr.cephadm_services[daemon_type_to_service(daemon_spec.daemon_type)].generate_config(
-                daemon_spec)
-
```

Comment:
You opted for calling `generate_config` from within `prepare_create()`. This is perfectly fine!

The problem with this change is: We have code paths into `_create_daemon` that don't call `prepare_create`, like:

https://github.com/ceph/ceph/blob/4db6aebee01d1499878f2cd776aaea027105d2f9/src/pybind/mgr/cephadm/module.py#L1600-L1616

https://github.com/ceph/ceph/blob/4db6aebee01d1499878f2cd776aaea027105d2f9/src/pybind/mgr/cephadm/module.py#L1957-L1960

which unlike `_apply_service()` do not call `prepare_create()`

https://github.com/ceph/ceph/blob/4db6aebee01d1499878f2cd776aaea027105d2f9/src/pybind/mgr/cephadm/serve.py#L560-L562

And that's the problem in https://tracker.ceph.com/issues/48373 . We have to make sure that all code paths into `_create_daemon` properly call `prepare_create()`. (or simply call `prepare_create` at this point here)

Reply:
@sebastian-philipp I can see 4 places we call `_create_daemon`. In `_apply_service` in serve.py which already calls `prepare_create`, once in osd.py where `prepare_create` isn't needed because it's making an osd,  in `_daemon_action` in cephadm/module.py which is covered in this commit as part of this PR https://github.com/ceph/ceph/pull/38549/commits/c0353e0c1d1a2580cee358774ad6e9a9f7739829 and in `_create_daemons` in cephadm/module.py which is covered by the call to `create_func` which should hold the `prepare_create` function for whatever daemon type is being created. Am I missing any?

Reply:
> Am I missing any? 

I think you're correct